### PR TITLE
Fix Android build jobs

### PR DESCRIPTION
**Pull request by gnzlbg**

### CODE REVIEW

**Comment by mati865** [comment]
For
```
Error occurred during initialization of boot layer
java.lang.module.FindException: Module java.se.ee not found
```
you should replace https://github.com/rust-lang/libc/blob/4ac8346b74f530d4573627fdd24b2fbe1018a3f2/ci/docker/arm-linux-androideabi/Dockerfile#L12 with `openjdk-8-jdk`.

**Comment by gnzlbg** [comment]
That used to work with the old NDK 15, but if interpreted the errors from the previous build correctly, doesn't work with the NDK19c. AFAICT, downgrading this to openjdk 10 should work, but should not work with the ndk20, which is in beta, so... i'm trying to see if there is a way to make this work with openjdk 11 and if not, i'll downgrade java till it works.

**Comment by mati865** [comment]
`java.se.ee` was deprecated in Java 9 and 10 to be finally removed in 11.
Ubuntu 19.04 repository contains OpenJDK 8, 11, 12 and 13.

**Comment by gnzlbg** [comment]
Oh, damn.

File: ci/android-install-sdk.sh [hunk 1]
```diff
@@ -49,8 +49,8 @@ esac;
 
 # Try to fix warning about missing file.
 # See https://askubuntu.com/a/1078784
-mkdir -p .android
-touch .android/repositories.cfg
+mkdir ./.android/
+touch ./.android/repositories.cfg
```
Comment (mati865) [inline ci/android-install-sdk.sh]: `touch` should re redundant because of `echo` below and the path should be probably `/root/.android`.

Comment (gnzlbg) [reply ci/android-install-sdk.sh]: Tried that (https://github.com/rust-lang/libc/pull/1344/commits/6b432e4a7d2a3f05414aa0a51bb3c53a9cf626f0) did not work.

Comment (mati865) [reply ci/android-install-sdk.sh]: Are you sure that directory was present?

File: ci/android-install-sdk.sh [hunk 2]
```diff
@@ -49,8 +49,7 @@ esac;
 
 # Try to fix warning about missing file.
 # See https://askubuntu.com/a/1078784
-mkdir ./.android/
-touch ./.android/repositories.cfg
+mkdir -p ./.android/
```
Comment (mati865) [inline ci/android-install-sdk.sh]: It will create `/.android` instead of `/root/.android`, docker containers pwd is usually set to `/`.

**Comment by gnzlbg** [comment]
Progress, now:

> The job exceeded the maximum log length, and has been terminated.

**Comment by mati865** [comment]
I'd rather silence `unzip` by adding `-q` but whatever.

**Comment by gnzlbg** [comment]
@mati865 we can do that to, `unzip` does dump quite a bit of output, but `sdkmanager` dumps 90% of the raw log file (check the raw output).

**Comment by mati865** [comment]
It didn't show on the website but you are right that progress bar was horrible.

**Comment by gnzlbg** [comment]
So now the x86 and x86_64 targets are green, but the arm and aarch64 targets fail to link:

```
= note: /android/ndk-arm/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot open crtbegin_dynamic.o: No such file or directory
          /android/ndk-arm/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot open crtend_android.o: No such file or directory
          /android/ndk-arm/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -llog
          external/jemalloc/src/jemalloc.c:1422: error: undefined reference to 'pthread_atfork'
          external/jemalloc/src/jemalloc.c:1422: error: undefined reference to 'pthread_atfork'
          external/jemalloc/src/jemalloc.c:1422: error: undefined reference to 'pthread_atfork'
          external/jemalloc/src/jemalloc.c:1422: error: undefined reference to 'pthread_atfork'
          external/jemalloc/src/jemalloc.c:1358: error: undefined reference to 'atexit'
          clang80: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by gnzlbg** [comment]
Also, this error:

```
cannot find crtbegin_dynamic.o: No such file or directory
          clang80: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by mati865** [comment]
I think you might have picked unfortunate API version for arm and aarch64:

<details>
<summary>find missing numbers :laughing: </summary>

```
$ find . -name "crtbegin_dynamic.o" | sort 
./platforms/android-16/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-16/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-17/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-17/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-18/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-18/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-19/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-19/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-21/arch-arm64/usr/lib/crtbegin_dynamic.o
./platforms/android-21/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-21/arch-x86_64/usr/lib64/crtbegin_dynamic.o
./platforms/android-21/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-22/arch-arm64/usr/lib/crtbegin_dynamic.o
./platforms/android-22/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-22/arch-x86_64/usr/lib64/crtbegin_dynamic.o
./platforms/android-22/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-23/arch-arm64/usr/lib/crtbegin_dynamic.o
./platforms/android-23/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-23/arch-x86_64/usr/lib64/crtbegin_dynamic.o
./platforms/android-23/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-24/arch-arm64/usr/lib/crtbegin_dynamic.o
./platforms/android-24/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-24/arch-x86_64/usr/lib64/crtbegin_dynamic.o
./platforms/android-24/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-26/arch-arm64/usr/lib/crtbegin_dynamic.o
./platforms/android-26/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-26/arch-x86_64/usr/lib64/crtbegin_dynamic.o
./platforms/android-26/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-27/arch-arm64/usr/lib/crtbegin_dynamic.o
./platforms/android-27/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-27/arch-x86_64/usr/lib64/crtbegin_dynamic.o
./platforms/android-27/arch-x86/usr/lib/crtbegin_dynamic.o
./platforms/android-28/arch-arm64/usr/lib/crtbegin_dynamic.o
./platforms/android-28/arch-arm/usr/lib/crtbegin_dynamic.o
./platforms/android-28/arch-x86_64/usr/lib64/crtbegin_dynamic.o
./platforms/android-28/arch-x86/usr/lib/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/21/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/22/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/23/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/24/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/26/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/27/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/28/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/16/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/17/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/18/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/19/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/21/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/22/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/23/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/24/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/26/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/27/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/28/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/16/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/17/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/18/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/19/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/21/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/22/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/23/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/24/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/26/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/27/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/i686-linux-android/28/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/21/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/22/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/23/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/24/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/26/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/27/crtbegin_dynamic.o
./toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/x86_64-linux-android/28/crtbegin_dynamic.o
```
</details>

**Comment by gnzlbg** [comment]
@mati865, we have to pick an NDK that has these, for which also a corresponding SDK image exists so that we can run the tests in the emulator, but there appear to be no SDK images for APIs > 25 for arm64 = / I don't recall if there were images for API == 24, but we need to find the intersection of NDK version that links properly and SDK with an emulator image for it.

**Comment by gnzlbg** [comment]
So @mati865 the build is now green with API 24. I've started cleaning up the build.rs for Android here, and I've run into `ENOATTR`. IIUC the motivation for adding it here was that #1030 it was available, as opposed as in Linux, and therefore the workaround wasn't necessary. But then... test failed, making the workaround necessary, yet the PR was merged anyways.

**Comment by mati865** [comment]
Yeah it's absent in NDK. Somehow it worked in toolchain I used back then, sorry for that.

File: src/unix/notbsd/linux/mod.rs [hunk 3]
```diff
@@ -1409,9 +1409,10 @@ pub const FALLOC_FL_ZERO_RANGE: ::c_int = 0x10;
 pub const FALLOC_FL_INSERT_RANGE: ::c_int = 0x20;
 pub const FALLOC_FL_UNSHARE_RANGE: ::c_int = 0x40;
 
-// On Linux, libc doesn't define this constant, libattr does instead.
-// We still define it for Linux as it's defined by libc on other platforms,
-// and it's mentioned in the man pages for getxattr and setxattr.
+#[deprecated(
+    since = "0.2.55",
+    note = "ENOATTR is not available on Android; use ENODATA instead"
```
Comment (mati865) [inline src/unix/notbsd/linux/mod.rs]: `s/Android/Linux/`

**Comment by mati865** [comment]
```
/checkout/target/aarch64-linux-android/debug/build/libc-test-28dc7719767fbc7c/out/linux_fcntl.rs:485: undefined reference to `__xpg_strerror_r'
/checkout/target/aarch64-linux-android/debug/build/libc-test-28dc7719767fbc7c/out/linux_fcntl.rs:485: undefined reference to `__xpg_strerror_r'
```
`nm` output of libc:
```
strerror_r.o:
0000000000000000 T __gnu_strerror_r
0000000000000000 T strerror_r
```

**Comment by gnzlbg** [comment]
So with this last commit all the low-hanging fixable fruit has been fixed. What remains is trickier to solve, has larger chances of breaking current users, or should be solved for multiple targets at once. A couple of issues have been opened to track that. 

I'll re-base, re-enable CI for all other targets, and try to land this without any more major changes.

**Comment by gnzlbg** [comment]
let's land this

@bors: r+

**Pull request merged**